format/format.go

```diff
@@ -0,0 +1,406 @@
+
+		if !hasErrors {
+			// record that the file was formatted
+			c.stats.Add(stats.Formatted, len(batch))
+		}
+
+		// create a release context
+		releaseCtx := walk.SetNoCache(ctx, hasErrors)
+
+		// post-processing
+		for _, file := range batch {
+			// check if the file has changed
+			changed, newInfo, err := file.Stat()
+			if err != nil {
+				return err
+			}
+
+			if changed {
+				// record that a change in the underlying file occurred
+				c.stats.Add(stats.Changed, 1)
+
+				log.Log(
+					c.changeLevel, "file has changed",
+					"path", file.RelPath,
+					"prev_size", file.Info.Size(),
+					"prev_mod_time", file.Info.ModTime().Truncate(time.Second),
+					"current_size", newInfo.Size(),
+					"current_mod_time", newInfo.ModTime().Truncate(time.Second),
+				)
+
+				// update the file info
+				file.Info = newInfo
+			}
+
+			// if this batch was not successful, we do not want to update the cache
+			// this ensures we will try to format these files again on the next invocation
+			if err := file.Release(releaseCtx); err != nil {
+				return fmt.Errorf("failed to release file: %w", err)
+			}
+		}
+
+		return nil
+	})
+}
+
+// match filters the file against global excludes and returns a list of formatters that want to process the file.
+func (c *CompositeFormatter) match(file *walk.File) []*Formatter {
+	// first check if this file has been globally excluded
+	if pathMatches(file.RelPath, c.globalExcludes) {
+		log.Debugf("path matched global excludes: %s", file.RelPath)
+
+		return nil
+	}
+
+	// a list of formatters that match this file
+	var matches []*Formatter
+
+	// otherwise, check if any formatters are interested in it
+	for _, formatter := range c.formatters {
+		if formatter.Wants(file) {
+			matches = append(matches, formatter)
+		}
+	}
+
+	return matches
+}
+
+// Apply applies the configured formatters to the given files.
+func (c *CompositeFormatter) Apply(ctx context.Context, files []*walk.File) error {
+	var toRelease []*walk.File
+
+	for _, file := range files {
+		matches := c.match(file) // match the file against the formatters
+
+		// check if there were no matches
+		if len(matches) == 0 {
+			// log that there was no match, exiting with an error if the unmatched level was set to fatal
+			if c.unmatchedLevel == log.FatalLevel {
+				return fmt.Errorf("no formatter for path: %s", file.RelPath)
+			}
+
+			log.Logf(c.unmatchedLevel, "no formatter for path: %s", file.RelPath)
+
+			// no further processing to be done, append to the release list
+			toRelease = append(toRelease, file)
+
+			// continue to the next file
+			continue
+		}
+
+		// record there was a match
+		c.stats.Add(stats.Matched, 1)
+
+		// check if the file is new or has changed when compared to the cache entry
+		if file.Cache == nil || file.Cache.HasChanged(file.Info) {
+			// add this file to a batch and if it's full, apply formatters to the batch
+			if key, batch := c.batches.Append(file, matches); len(batch) == c.batchSize {
```

Comment:
Whenever I see a pattern like this, I immediately look for the code after the loop that's supposed to handle the last batch. I don't see that here. Is it possible there's an incomplete batch at the end that we don't format?

Edit: I see the `Close` method below, it looks like it's designed to handle this last batch scenario. I don't understand how it gets called, though.

Reply:
Got it! In that case, it smells a little odd to me that the "process full batches" vs "process the last incomplete batch" code lives so far apart (and at different "layers" (whatever the difference between `cmd/format/format.go` vs `format/format.go` is) of the stack?)

That said, it sounds like this works. Hopefully it's tested, and maybe someday in the future (when I'm more comfortable with go and treefmt, I'll have thoughts about if this could be written differently).

I'm going to mark this resolved.